Feature/3732 trashed comments handling

mzorz opened the pull request:
Addresses https://github.com/wordpress-mobile/WordPress-Android/issues/3732

### Notes:
- I’m using the same Approve icon for Restore (only changed the name) - design team might want to take a look @mattmiklic (check screenshot below)
- Saw the “Not spam” (unspam) button in the app already sets the comment as `APPROVED` when clicked, so doing the same for the “Restore” action
- There didn’t seem to be a `DELETE PERMANENTLY` endpoint in XMLRPC API as per the documentation, and while I’ve seen the WP API supports this by having the `$force_delete` param set to true, it wasn’t there in the XMLRPC API docs, but I tried nevertheless and sending a fifth boolean parameter to the XMLRPC API  on the `wp.deleteComment` method works like as expected actually deleting the comment.

Actions available on an item in the  `Trashed` list are:

- a) Restore (which is APPROVE)
- b) Delete (permanently)
- c) SPAM

If you hit Delete, we now display a dialog to confirm the action.

Similarly, the list selection has been implemented following this line. See long press on a comment in the filtered list for this.

## Screenshots
### Individual comment
![screenshot_2016-02-12-18-20-31](https://cloud.githubusercontent.com/assets/6597771/13020690/959e478e-d1b5-11e5-98de-8a6970f349ca.png)

![screenshot_2016-02-12-18-20-34](https://cloud.githubusercontent.com/assets/6597771/13020705/ab14c34a-d1b5-11e5-9374-be49d3df0a37.png)


### Multiple selection
![screenshot_2016-02-12-18-20-48](https://cloud.githubusercontent.com/assets/6597771/13020729/ca40ca66-d1b5-11e5-9198-4d5e4ed0a2d4.png)
![screenshot_2016-02-12-18-20-51](https://cloud.githubusercontent.com/assets/6597771/13020730/ca417948-d1b5-11e5-8c6a-dc16d762fbfa.png)
![screenshot_2016-02-12-18-20-58](https://cloud.githubusercontent.com/assets/6597771/13020728/ca40a568-d1b5-11e5-999d-de129cc70c98.png)

nbradbury commented:
If from the detail view I delete the only trashed comment, the list view is blank when I return to it (ie: there's no empty list message).

![delete1](https://cloud.githubusercontent.com/assets/3903757/13021546/dadcac84-d1a9-11e5-8532-4a58c1d15cc9.png)

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentActions.java:
```diff
@@ -437,7 +437,8 @@ public void run() {
                                         final Comment comment,
-                                        final CommentActionListener actionListener) {
+                                        final CommentActionListener actionListener,
+                                        final boolean forceDelete) {
```
This feels nitpicky, but I think the `forceDelete` parameter should be named `deletePermanently` for clarity.

mzorz replied:
:100: was trying to follow along the names used on the XMLRPC API but it makes perfect sense from the CommentActions class point of view :+1:

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java:
```diff
@@ -282,7 +284,31 @@ public void onClick(View v) {
             @Override
             public void onClick(View v) {
-                moderateComment(CommentStatus.TRASH);
+                if (mComment.getStatusEnum() == CommentStatus.TRASH) {
```
This would benefit from a comment stating that the dialog is necessary because trashing an already trashed comment deletes it permanently.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java:
```diff
@@ -282,7 +284,31 @@ public void onClick(View v) {
+                            });
+                    dialogBuilder.setNegativeButton(
+                            getResources().getText(R.string.no),
+                            new DialogInterface.OnClickListener() {
```
Just an FYI: you can pass `null` here since the click listener does nothing (something I didn't know until relatively recently).

mzorz replied:
:+1: I tend to follow the formats / look alike of the code I'm handed only, sometimes it seems the listeners are set on purpose for later doing something else (i.e. UI analytics tracking?)

nbradbury commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -273,6 +273,7 @@
     <string name="comment_status_unapproved">Pending</string>
     <string name="comment_status_spam">Spam</string>
     <string name="comment_status_trash">Trashed</string>
+    <string name="comment_status_delete">DELETE</string>
```
This string doesn't appear to be used anywhere.

mzorz replied:
:+1:

nbradbury commented:
When a single comment is selected in the list and you choose to delete it, the wording on the dialog is still plural ("these comments").

![delete1](https://cloud.githubusercontent.com/assets/3903757/13022727/4e29b572-d1b1-11e5-9f50-9f1a25c41918.png)

nbradbury commented:
I'm a bit bothered by the number of places which check for an existing `TRASH` status to determine whether trashing will permanently delete. It all works well, but it's not very self-documenting. Perhaps we could add a routine to do this check instead? For example:

```
boolean willTrashingPermanentlyDelete(Comment comment) {
	return comment.getStatusEnum() == CommentStatus.TRASH;
}
```

Related to this is the fact that I just discovered trashing a spam comment permanently deletes it, so if we rely on this routine we could simply check for `CommentStatus.SPAM` as well.

mattmiklic commented:
I missed the question about the arrow icon before the menu was merged in, but I have a question about how it's implemented. The bottom-right-pointing arrow looks standard for the "spinner" control I see in the Android developer documentation. The design actually uses what the Material Design guidelines call a "[dropdown button](https://www.google.com/design/spec/components/buttons.html#buttons-persistent-footer-buttons)", though. I'd assumed that that downward-pointing arrow included in the Material Design guidelines was a standard system element. I'm having some trouble finding technical documentation on anything called a "dropdown button," though.

For the Restore action, we should use the gridicons-undo icon. I'd like to eventually start exporting vector icons for Android but for the sake of expediency I've exported the icon as a PNG, matching the color and size of the existing Approve icon (32dp square). https://cloudup.com/iPoCnHwjRCi

I'm not sure about the design of the multi-selection mode -- what I understand in the Material Design guidelines states that checkboxes [should not be used for in multi-selection mode](https://www.google.com/design/spec/patterns/selection.html):

> Checkboxes, or custom analogs, should only be used to indicate completion, such as checking items off task lists or checklists.
>
> They should not be used for selection on mobile.

I think simply changing the background color of the selected item's row is sufficient. In Google's own apps, and in our Hold for Moderation/Blacklist options in Site Settings, the toolbar color changes when you enter multi-selection mode. I'm not sure how that would work since we have our extra bar along the bottom.

nbradbury commented:
> I'm not sure about the design of the multi-selection mode -- what I understand in the Material Design guidelines states that checkboxes should not be used for in multi-selection mode:

I should mention that the multi-select behavior was something I added a while ago, so it's not specific to these two PRs. We can file a separate issue to change it. Funny enough, though, I modeled it on the Gmail app, which still uses checkboxes in multi-select :unamused:

nbradbury commented:
@mattmiklic Thinking about multi-select further, I'm not convinced our implementation violates the guidelines. I think the guidelines are saying not to use checkbox [components](https://www.google.com/design/spec/components/selection-controls.html#selection-controls-checkbox) to indicate selection. This is something pre-ICS apps would often do, [like this](https://cloudup.com/cG40dk_Jyqb).

Like both Gmail and Google Inbox, we're using large checkbox images to indicate selection as an additional cue beyond changing the background color, and I think we're ok to do that.

mzorz commented:
Thanks for your comments guys!

I think I’ve addressed all of them and also did some more tests, here my notes on the work done:

### Code and general

- Regarding empty views: checked that and corrected, I think for all the cases now (no need to make a separate issue :+1:)

- In checking this (and thank you @nbradbury to make me look a bit deeper), I also found a bug: when I trash an item, and I quickly change filter before the UNDO snackbar is gone, then move to filter by Trashed, then click on the trashed item, the Comment Detail view is shown but all empty. If you go back, reload comments (by refiltering or refreshing) and do it again, detail shows up correctly. This was because the CommentDetailFragment took its data from the local database, but the database has not yet been updated at the time the fragment was being created.Fixed this as well. Root cause is trashed items would be deleted from the database earlier (before filtering by type was implemented), now they are deleted only if `deletePermanently==true` (as TRASH actually means something else now). 

- another bug: if items of a given type would come null/empty from the server, but we had for some reason still at least one item of that type in the db, that item would still be shown on the screen (it got never deleted). Root cause was we only updated the database when the server returned something, but didn’t update it otherwise. Now we also delete items on a specific filter if server returns an empty array successfully for any given filtered request.

- also, if a comment was deleted on the server, but the android app had a list of items, as they were only treated with INSERT OR REPLACE it wouldn’t ever update the deleted comment. Fixed that now implementing an interface `DatabasePersistCallback` that can be passed to `ApiHelper.refreshComments()`. Felt the best way to separate responsibility here as we didn’t want to save data right away anymore but to have someone else handle this.

- There can always be a blind spot when you trash an element on say, the Pending/approved/all lists, and immediately after go to check the Trash list, which is as long as `the time taken by the snackbar plus the time to make the request to the server`.

All these 4 inter-related changes are on commit https://github.com/wordpress-mobile/WordPress-Android/commit/214708fad10f8396fb0cab82cc35e5aa9750a759

Also:

- Changed plural/singular on lists depending on whether one or more items are selected.

- Good catch about the SPAM items being directly deleted from server when trashed. And added the suggested `willTrashingPermanentlyDelete` solution as a method to Comment object as well :+1:. I ended up using it only in once place as I needed to make a distinction between SPAM or TRASH’d items elsewhere but good to have a clearer condition check there with the use of that naming on that method.



### Design

Thanks for the restore icon @mattmiklic, I’m using it now and renamed to `ic_action_restore` to follow the naming convention the other action assets seemed to have within the Android project.

Re: multiselection, I agree with @nbradbury about the multi-selection UX, it’s on par with latest practices

Regarding the icon for drop downs, it’s ok to just leave the one being used on spinners anywhere else. What I was saying in my comment in the other issue is that, in case we really want to use the pointing-down-triangle icon, I’d need it to be resolved with 9-patch to use it in the same manner that the current Spinner is implemented. I can replace the asset if provided.


### Other

BTW I raise a question: if we’re ok with the hide/show on scroll down/up behaviour on Comments, we should consider bringing the same UX / functionality to the Stats screen, what do you guys think @mattmiklic @nbradbury?

nbradbury commented:
> if we’re ok with the hide/show on scroll down/up behaviour on Comments, we should consider bringing the same UX / functionality to the Stats screen

Yes, I think that makes sense, and possibly the reader as well.

mzorz commented:
> >  if we’re ok with the hide/show on scroll down/up behaviour on Comments, we should consider bringing the same UX / functionality to the Stats screen

> Yes, I think that makes sense, and possibly the reader as well.

Created issue https://github.com/wordpress-mobile/WordPress-Android/issues/3750

nbradbury commented:
> For Comments, when the filter is switched we need this...But for the ApiHelper.RefreshBlogContentTask we need a slightly different one

Ah, I hadn't noticed that. In that case your approach makes more sense.

mzorz commented:
:+1: are we good to merge then?

nbradbury commented:
Just noticed this while testing a self-hosted site. If I'm viewing all comments and I mark the only comment as spam, the empty message says "No all comments" instead of simply "No comments."

![no-all-comments](https://cloud.githubusercontent.com/assets/3903757/13056353/d08fda7c-d3e2-11e5-9972-5ffc8dac4b0f.png)

mzorz commented:
Awesome catch - will check in a min

mzorz commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentsListFragment.java:
```diff
@@ -222,7 +224,8 @@ public FilterCriteria onRecallSelection() {
                 if (emptyViewMsgType == EmptyViewMessageType.NO_CONTENT) {
-                    if (mFilteredCommentsView.getCurrentFilter() == null) {
+                    FilterCriteria filter = mFilteredCommentsView.getCurrentFilter();
+                    if (filter == null || CommentStatus.UNKNOWN.equals(filter)) {
```
Here the culprit for that awkward  "No all comments"  message @nbradbury :)

nbradbury replied:
> Here the culprit for that awkward "No all comments" message

Yep, that fixed it!

nbradbury commented:
:shipit:
